kong/tools/stream_api.lua

```diff
@@ -0,0 +1,100 @@
+
+require "lua_pack"
+
+
+local kong       = kong
+local st_pack    = string.pack      -- luacheck: ignore string
+local st_unpack  = string.unpack    -- luacheck: ignore string
+local st_format  = string.format
+
+local PREFIX = ngx.config.prefix()
+
+local stream_api = {}
+
+local _handlers  = {}
+
+
+function stream_api.load_handlers()
+  local utils = require "kong.tools.utils"
+
+  for plugin_name in pairs(kong.configuration.loaded_plugins) do
+    local loaded, custom_endpoints = utils.load_module_if_exists("kong.plugins." .. plugin_name .. ".api")
+    if loaded and custom_endpoints._stream then
+      kong.log.debug("Register stream api for plugin: ", plugin_name)
+      _handlers[plugin_name] = custom_endpoints._stream
+      custom_endpoints._stream = nil
+    end
+  end
+end
+
+function stream_api.request(key, data, socket_path)
+  if type(key) ~= "string" or type(data) ~= "string" then
+    kong.log.error("Key and data must be strings")
+    return
+  end
+  if #data > 8000 then
```

Comment:
yep, magic numbers are ugly.   i'm keeping it at 8000, because of framing, and using [8180-len(plugin_name)] for requests and 8182 for responses seems silly.